### ISSUE TITLE
Set memory limit for preloading

### ISSUE THREAD

**Issue by raziel057** [symfony/symfony#45950]
### Description

Hello,

I currently encounter an issue related to the fact the amount of files to preload at server start exceed the memory limit fixed to 128M in php.ini. I'm not sure why I got this issue after an upgrade from SF 4.4 / PHP 7.4 to SF 6.0 / PHP 8.1 (I fixed all deprecations) as it was not the case before, but in any case, to avoid this kind of issue what do you think about adding a ``ini_set('memory_limit, '-1');`` on top of the ``AppKernelProdContainer.preload.php`` generated file?

Or maybe a better idea would be to let the user override the default value if needed in config. In that case, the ini_set instruction would be dumped in the preload file with the user defined value, only if a parameter like ``preloading.memory_limit`` have been provided.

@nicolas-grekas What do you think about this?

It's strange that we can't just set a different memory_limit for preloading step.

**Comment by PabloKowalczyk** [comment]
I also encountered this bug(?), possible workaround is to set higher memory limit in `php.ini` and then reduce it in `index.php` - this is what I do.

**Comment by raziel057** [comment]
@PabloKowalczyk Another workaround could be to require another app specific preload file ``MyAppKernelProdContainer.preload.php`` that itself require the AppKernelProdContainer.preload.php file but it would be nice to have an embedded solution as soon as multiple people can be concerned by this issue.

**Comment by michaljusiega** [comment]
Just wondering only, because the problem with `memory_limit` occurs more often in PHP community.
We live in 2022, people uses more complex applications but still using 128MB memory per script.

How hard is try to increase `memory_limit` parameter e.g twice ? This is not any issue or bug, current applications really require a lot of hardware resources. It's just like that you want to play the new released game on old computer. Theoretically impossible on full details ;)

Sometimes upgrades of softwares requires also upgrade of hardware IMO.

**Comment by raziel057** [comment]
@michaljusiega The fact is that lots of even complex application don’t need more than 128M per request. The preloading phase is only run at server startup and more memory consuming. It seems better for security and code quality to be able to allocate more memory for a specific process than globally while not needed.

**Comment by PabloKowalczyk** [comment]
@michaljusiega image you have server with 8GB RAM, with 256MB memory limit (per process) in PHP(-FPM) you can safely create 14 workers, it's good number for small site, but too low for bigger apps. In my current project I set memory limit for 22MB, even in 2022 :)

**Comment by gggeek** [comment]
Dunno.
I prefer to limit the rate of incoming requests as close to upstream as possible, to prevent overloading any of the services downstream. That means in most cases via the webserver or reverse-proxy/cdn config rather than php memory limit.
Fine-grained limits on php memory is always a complex thing - one often has web pages for admins who want to upload huge images, migration scripts which massage half the database and similar, intermingled with web pages which logically should only require a couple of megs.
The only use for memory_limit that I can think of is badly coded apps which have runaway loops and allocations. But those are easy to spot and fix in any case, as they will overflow even a 4GB limit.
Long story short: I always set php memory limit to 2GB. And never had a site crash or slow to a crawl because of that - whenever resources are exhausted because of web traffic, it is either php melting the cpu, or mysq/solr/redis/glusterfs eating all the available memory...

**Comment by raziel057** [comment]
In fact I didn't know but it's already recommended to not target directly the preloading file in cache.
See https://symfony.com/doc/current/performance.html#use-the-opcache-class-preloading

Which is also in symfony demo app:
https://github.com/symfony/demo/blob/main/config/preload.php

In that case, the proposed workaround https://github.com/symfony/symfony/issues/45950#issuecomment-1090401227 seems a good solution. We just have to set the ini_set('memory_limit', 'xx') in the project preloading file.

So no need to modify the PhpDumper.

**Comment by PabloKowalczyk** [comment]
@raziel057 nice tip!

My `preload.php` script:

```php
<?php

declare(strict_types=1);

$path = \dirname(__DIR__) . '/var/cache/prod/Zwzt_KernelProdContainer.preload.php';
if (\file_exists($path)) {
    $originalMemoryLimit = \ini_get('memory_limit');
    // Temporary increase memory limit
    \ini_set('memory_limit', '128M');
    require $path;
    // Restore original memory limit
    \ini_set('memory_limit', $originalMemoryLimit);
}
```

**Comment by raziel057** [comment]
@PabloKowalczyk no need to reset the original memory limit as the preloading file is only required at server startup. All requests are constraints to the predefined memory limit (from PHP config)

**Comment by PabloKowalczyk** [comment]
> @PabloKowalczyk no need to reset the original memory limit as the preloading file is only required at server startup. All requests are constraints to the predefined memory limit (from PHP config)

Nice, even better. Thanks.

**Comment by 7-zete-7** [comment]
Good to find this thread.

Thanks @raziel057 and @PabloKowalczyk for your research!

The final content of `config/preload.php` is:
```php
<?php

declare(strict_types=1);

if (file_exists(dirname(__DIR__).'/var/cache/prod/App_KernelProdContainer.preload.php')) {
    ini_set('memory_limit', '256M'); // set value you need here

    require dirname(__DIR__).'/var/cache/prod/App_KernelProdContainer.preload.php';
}
```

I also tried to use environment variables to configure this value, but Opcache do not forwards environment variables in to preload script.
